Triangulation_2/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h

```diff
@@ -171,8 +162,11 @@ class Polyline_constraint_hierarchy_2
   typedef typename Context_list::iterator Context_iterator;
 
   typedef std::set<Constraint_id>           Constraint_set;
-  typedef std::map<Edge, Context_list*,
-                   Pair_compare>            Sc_to_c_map;
+#if CGAL_USE_BOOST_UNORDERED
+  typedef boost::unordered_flat_map<Edge, Context_list*, boost::hash<Edge>> Sc_to_c_map;
+#else
+  typedef std::unordered_map<Edge, Context_list*, boost::hash<Edge>> Sc_to_c_map;
+#endif
```

Comment:
## Order of the sequence of values from the range `CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`

@janetournois @afabri

- The patch https://github.com/CGAL/cgal/pull/8273/commits/93fd96644c9e2a1aab33433880afada3e4eed02c makes the range [`CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`](https://doc.cgal.org/6.0/Triangulation_2/classCGAL_1_1Constrained__triangulation__plus__2.html#af25114a7e1675194367f8f9de9de90d2) non-deterministic: instead of a `std::map` with a custom `compare` functor, now the sub-constraints are in an *unordered* map, for efficiency reasons. Note that the range was never documented as sorted or deterministic.
- As my patch had broken a test from Mesh_2 (the test `test_meshing`), I had to fix that non-determinism. That was complicated to fix directly in `Constrained_triangulation_plus_2` or `Constrained_triangulation_plus_2`. Instead, I have chosen to fix the only place in CGAL where the order of that range mattered, in `Mesh_2/include/CGAL/Mesh_2/Refine_edges.h`. See the commit https://github.com/CGAL/cgal/pull/8273/commits/82b53596fd6beaaf6c5accf2182cbdd3553b3365.

Is that correct? That is a sort of breaking change, but the ordering of `Ctp_2::constraints()` was never documented.

Reply:
Jane (IRL) suggested that:

  -  it is documented that there is a breaking change
  -  and add a Tag to `Constrained_triangulation_plus_2` to allow users to fallback to the previous behavior.

Reply:
(I have copy-pasted my previous *comment* into a *discussion*. So that all the messages about this discussion are at the same place.)